AutoImplement/DecoratorBuilder.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace AutoImplement {
+   public class DecoratorBuilder : Builder {
+      private readonly List<string> implementedMethods = new List<string>();
+
+      private string innerObject;
+
+      protected override string ClassNamePostfix => "Decorator";
+
+      protected override void IncludeExtraMembers(Type type) {
+         base.IncludeExtraMembers(type);
+         innerObject = GetInnerObjectName(type);
+         var typeName = type.CreateCsName(type.Namespace);
+         AppendLine($"protected {typeName} {innerObject} {{ get; set; }}");
+      }
+
+      protected override void AppendMethod(MethodInfo info, MemberMetadata method) {
+         var returnClause = method.ReturnType == "void" ? string.Empty : "return ";
+
+         // Use an explicit implementation only if the signature has already been used
+         // example: IEnumerable<T>, which extends IEnumerable
+         // since expicit implementation can't be virtual, have the explicit one call the normal one.
+         // In the case of Enumerable, this is correct. Hopefully there aren't too many interfaces that makes this happen, cuz it's WEIRD.
+         // When it does happen, hopefully the creator of the child interface wanted the child method to behave as the parent method... in which case this is the correct implementation.
+         if (implementedMethods.Any(name => name == $"{method.Name}({method.ParameterTypes})")) {
+            AppendLine($"{method.ReturnType} {method.DeclaringType}.{method.Name}({method.ParameterTypesAndNames})");
+            using (Indent()) {
+               AppendLine($"{returnClause}{method.Name}({method.ParameterNames});");
+            }
+            return;
+         }
+
+         AppendLine($"public virtual {method.ReturnType} {method.Name}({method.ParameterTypesAndNames})");
+         using (Indent()) {
+            AssignDefaultValuesToOutParameters(info);
+            AppendLine($"if ({innerObject} != null)");
+            using (Indent()) {
+               AppendLine($"{returnClause}{innerObject}.{method.Name}({method.ParameterNames});");
+            }
+            if (returnClause != string.Empty) {
+               AppendLine("else");
+               using (Indent()) {
+                  AppendLine($"return default({method.ReturnType});");
+               }
+            }
+         }
+
+         implementedMethods.Add($"{method.Name}({method.ParameterTypes})");
+      }
+
+      protected override void AppendProperty(PropertyInfo info, MemberMetadata property) {

```

Comment:
These methods share like 85% common code - maybe consider using a helper method that takes in the property line, and the return default line

https://github.com/haven1433/autoimplement/blob/4dd8593c581ca48d68fa4c31a40b5153f9c68400/AutoImplement/DecoratorBuilder.cs#L21-L117